src/iceberg/parquet/parquet_schema_util.cc

```diff
@@ -30,8 +33,21 @@ Result<std::vector<int>> SelectedColumnIndices(const SchemaProjection& projectio
   return NotImplemented("NYI");
 }
 
-Result<bool> HasFieldIds(const ::parquet::schema::NodePtr& root_node) {
-  return NotImplemented("NYI");
+bool HasFieldIds(const ::parquet::schema::NodePtr& node) {
+  if (node->field_id() >= 0) {
+    return true;
+  }
+
+  if (node->is_group()) {
+    auto group_node = internal::checked_pointer_cast<::parquet::schema::GroupNode>(node);
+    for (int i = 0; i < group_node->field_count(); i++) {
+      if (HasFieldIds(group_node->field(i))) {
+        return true;
+      }
+    }
```

Comment:
What does this means? Any leaf has field-id?

Reply:
Just simply check if any field (whether leaf or non-leaf) has field-id. This is in sync with the Java impl.

Reply:
I think we cannot blindly check if all fields have ids because it is valid to have some fields that do not have ids, especially when variant type is supported.

Reply:
I knew that for 3-levels complex type, fieldId might not exists, but here I don't fully understand the purpose, so any leaf has field-id means writer is field-id aware?

Reply:
I think Iceberg only supports three-level list encoding per the spec: https://iceberg.apache.org/spec/#parquet

> Lists must use the [3-level representation](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#lists).